source/common/protobuf/deterministic_hash.cc

```diff
@@ -0,0 +1,239 @@
+    for (const auto q : reflection->GetRepeatedFieldRef<value_type>(message, field)) {             \
+      seed = HashUtil::xxHash64(absl::string_view{reinterpret_cast<const char*>(&q), sizeof(q)},   \
+                                seed);                                                             \
+    }                                                                                              \
+  } else {                                                                                         \
+    const auto q = reflection->Get##get_type(message, field);                                      \
+    seed =                                                                                         \
+        HashUtil::xxHash64(absl::string_view{reinterpret_cast<const char*>(&q), sizeof(q)}, seed); \
+  }
+
+uint64_t reflectionHashMessage(const Protobuf::Message& message, uint64_t seed = 0);
+uint64_t reflectionHashField(const Protobuf::Message& message,
+                             const Protobuf::FieldDescriptor* field, uint64_t seed);
+
+struct MapFieldComparer {
+  MapFieldComparer(const Protobuf::Message& first_msg)
+      : reflection_(*first_msg.GetReflection()), descriptor_(*first_msg.GetDescriptor()),
+        key_field_(*descriptor_.map_key()), compare_fn_(selectCompareFn()) {}
+  bool operator()(const Protobuf::Message& a, const Protobuf::Message& b) {
+    return (this->*compare_fn_)(a, b);
+  }
+
+private:
+  bool compareByInt32(const Protobuf::Message& a, const Protobuf::Message& b) {
+    return reflection_.GetInt32(a, &key_field_) < reflection_.GetInt32(b, &key_field_);
+  }
+  bool compareByUInt32(const Protobuf::Message& a, const Protobuf::Message& b) {
+    return reflection_.GetUInt32(a, &key_field_) < reflection_.GetUInt32(b, &key_field_);
+  }
+  bool compareByInt64(const Protobuf::Message& a, const Protobuf::Message& b) {
+    return reflection_.GetInt64(a, &key_field_) < reflection_.GetInt64(b, &key_field_);
+  }
+  bool compareByUInt64(const Protobuf::Message& a, const Protobuf::Message& b) {
+    return reflection_.GetUInt64(a, &key_field_) < reflection_.GetUInt64(b, &key_field_);
+  }
+  bool compareByBool(const Protobuf::Message& a, const Protobuf::Message& b) {
+    return reflection_.GetBool(a, &key_field_) < reflection_.GetBool(b, &key_field_);
+  }
+  bool compareByString(const Protobuf::Message& a, const Protobuf::Message& b) {
+    std::string scratch_a, scratch_b;
+    return reflection_.GetStringReference(a, &key_field_, &scratch_a) <
+           reflection_.GetStringReference(b, &key_field_, &scratch_b);
+  }
+  using CompareMemberFn = bool (MapFieldComparer::*)(const Protobuf::Message& a,
+                                                     const Protobuf::Message& b);
+  CompareMemberFn selectCompareFn() {
+    using Protobuf::FieldDescriptor;
+    switch (key_field_.cpp_type()) {
+    case FieldDescriptor::CPPTYPE_INT32:
+      return &MapFieldComparer::compareByInt32;
```

Comment:
I think you'd always need the switch because you can't template-resolve the "unknown" type.
Except in the one model I tried where you can have a `MapFieldComparator<T>` and offload the switch to a create-function, but as mentioned it turns out std::sort can't take a polymorphic type, so for performance it makes more sense to unify the switch into the monomorphic comparator's initializer, because wrapping a polymorphic in a monomorphic wrapper would end up having to double-de-reference every time.